### DISCUSSION TITLE
Out of memory

### DISCUSSION THREAD

**Discussion by RECHOA** [idaholab/moose#22452] [Q&A General] [answered]
When i run a case, i met a error shows 'EXIT CODE: 9'. Learning from other discussions, it means running out of memory. But acturally, the computational memory of our HPC has been added to 2TB, which is the maximum memory it could provide. Here is a part of the terminal log:
```
  Initializing
    Updating Because Mesh Changed
      Updating Mesh.                                                                     [[33m 11.78 s[39m] [[33m    8 MB[39m]
    Finished Updating Because Mesh Changed                                               [[33m 12.00 s[39m] [[33m   36 MB[39m]
.                                                                     [[33m 15.24 s[39m] [[33m    0 MB[39m]
    Finished Updating Because Mesh Changed                                               [[33m 15.55 s[39m] [[33m    0 MB[39m]
....
      Updating Geometric Search
        Updating Displaced GeometricSearch
          Finding Nearest Nodes
            Building Node To Elem Map                                                    [[33m  9.36 s[39m] [[33m  358 MB[39m]
          Finished Finding Nearest Nodes                                                 [[33m  9.37 s[39m] [[33m  358 MB[39m]
        Finished Updating Displaced GeometricSearch                                      [[33m  9.60 s[39m] [[33m  363 MB[39m]
      Finished Updating Geometric Search                                                 [[33m  9.60 s[39m] [[33m  363 MB[39m]
      Finished Building Node To Elem Map                                                 [[33m  9.33 s[39m] [[33m  426 MB[39m]
    Still Initializing Equation Systems.........                                         [[33m 96.82 s[39m] [[33m 1386 MB[39m]
    Initializing Displaced Equation System..........                                     [[33m 62.36 s[39m] [[33m  717 MB[39m]
      Finished Updating Mesh                                                             [[33m 11.17 s[39m] [[33m    3 MB[39m]
    Finished Updating Because Mesh Changed                                               [[33m 11.47 s[39m] [[33m    3 MB[39m]
  Finished Initializing                                                                  [[33m198.22 s[39m] [[33m 2143 MB[39m]
[DBG][ACT] TASK ([33m            check_output[39m) TYPE ([33m               CheckOutputAction[39m) NAME ([33m                [39m) Memory usage 7089MB
[DBG][ACT] TASK ([33m         check_integrity[39m) TYPE ([33m            CheckIntegrityAction[39m) NAME ([33m                [39m) Memory usage 7089MB
[DBG][ACT] TASK ([33m               no_action[39m) TYPE ([33m                     EmptyAction[39m) NAME ([33m    BEEsPressure[39m) Memory usage 7089MB
[DBG][ACT] TASK ([33mfinish_input_file_output[39m) TYPE ([33m                     EmptyAction[39m) NAME ([33m    BEEsPressure[39m) Memory usage 7089MB
[DBG][ACT] Finished executing all actions with memory usage 7089MB

Finished Setting Up                                                                      [[33m1289.15 s[39m] [[33m 7037 MB[39m]
Framework Information:
MOOSE Version:           git commit ccf25ce on 2022-07-13
LibMesh Version:         
PETSc Version:           3.16.5
SLEPc Version:           3.16.2
Current Time:            Thu Oct 20 17:57:41 2022
Executable Timestamp:    Sat Oct  8 02:20:19 2022

Parallelism:
  Num Processors:          60
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 1713784
    Local:                 30461
    Min/Max/Avg:           24821/30461/28563
  Elems:                   
    Total:                 4937408
    Local:                 82442
    Min/Max/Avg:           80273/88294/82290
  Num Subdomains:          3143
  Num Partitions:          60
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                6855136
  Num Local DOFs:          121844
  Variables:               { "temp" "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                63691744
  Num Local DOFs:          1068823
  Variables:               { "fast_neutron_fluence" "baf" "hoop_stress_bees" "hoop_strain_bees" } "rad_disp" 
                             "maxPrincipal_stress" { "creep_strain" "thermal_strain" "swell_strain" "Weibull_failure_probability" 
                             "Pdpenetration_depth" "qpoint_penetration" "paired_temp" } { "penetration" 
                             "contact_pressure" "nodal_area_pellet_clad_mechanical" } 
  Finite Element Types:    "MONOMIAL" "LAGRANGE" "MONOMIAL" "MONOMIAL" "LAGRANGE" 
  Approximation Orders:    "CONSTANT" "FIRST" "CONSTANT" "CONSTANT" "FIRST" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             FunctionDT
  Solver Mode:             Preconditioned JFNK
  MOOSE Preconditioner:    SMP

[31mLEGACY MODES ENABLED:[39m
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.
[39m

    Setting Up Materials
      Computing Initial Material Values.....
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 87204 RUNNING AT ga0304
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

Also, I add the
```
[Debug]
  show_actions = true
[]
```
to to see more information about the memory cost . 
The complete record of terminal log shows in the following file (result_log.txt):
[result_log.txt](https://github.com/idaholab/moose/files/9829129/result_log.txt)

The `Executioner `setting is
```
[Executioner]
  type = Transient
  petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = '201                hypre    boomeramg      4'
  line_search = 'none'
  solve_type = PJFNK
  l_tol =      0.0001
  l_max_its =  20
  nl_rel_tol = 1e-06
  nl_abs_tol = 1e-05
  nl_max_its = 15
  dtmin = 0.1
  dtmax = 1.0e6
  start_time = 0.0
  end_time = 9.0e7
  [./TimeStepper]
    type = FunctionDT
    function = dts
    min_dt = 1.0
  [../]
 []
```
I don't know which part of my case causing the memory exceed the maximum, and could anyone give better petsc options setting in the `Executioner `?

Thank you in advance.

**Comment by GiudGiud** [comment, marked as answer]
Hello

You're at least at 420Gb used.
Could you try distributed meshes (--distributed-mesh on the command line) to reduce the memory consumption?
Are you running this on a cluster scheduler that could have killed it for other reasons?

petsc options should not be an issue yet when computing materials

Guillaume

**Reply by RECHOA** [reply]
The computational memory of HPC has been added to 2TB, it is quite strange that this case failed but only 420Gb used.
The mesh file isn't created by using MOOSE built-in `mesh `block (acturally by Cubit), i'm not sure whether the `--distributed-mesh` could work, but i'll try it.
I run this case in HPC for many times but it failed due to the same error. Besides, for similar cases with less number of elements, they could run successfully.

**Reply by RECHOA** [reply]
I tried the `--distributed-mesh` command, but it got the same error. Besides, the memory usage  seems to have increased rather than decreased. Here is the new new_result_log.txt file:
[new_result_log.txt](https://github.com/idaholab/moose/files/9830831/new_result_log.txt)

Does the memory usage show in the file reperesent the total meomry or the memory in one processor?

**Reply by RECHOA** [reply]
I used the command `watch “free -g” `to see the change of the memory usage, and I noticed that the memory usage was normal (about 400~500GB) during the case running  until it run to the step:
```
  Setting Up Materials
     Computing Initial Material Values.....
```
at that time, the memory usage would increase dramatically (exceed 2TB), and then the case failed.

What should i do to get over this problem?

**Reply by GiudGiud** [reply]
The previous log was showing 7Gb used, this one shows 6167MB so it's a little better with distributed meshes.

Why do you have so many materials? Seems like thousands of materials and auxkernels. 
Why cant you use the same auxkernel or materials on all the blocks? What is different?
3000 subdomains is a little much as well.

**Reply by RECHOA** [reply]
There are a lot of subdomains (spheres) in my case. In each subdomain (sphere), its eigen strain is related to its location, since spherical coordinate system was created in every sphere. So there're many materials.

**Reply by GiudGiud** [reply]
cant it be the same material defined across multiple blocks?
The location is a parameter of the material?

**Reply by RECHOA** [reply]
yes, when calcuting the eigen strain for each sphere, the location is a parameter.

**Reply by GiudGiud** [reply]
I'd change that to use the same material everywhere, and have the location parameter be either:
- a really long vector (temporarily)
- read from a text file
Keeping the number of objects down will help tremendously with simulation time and memory cost.

**Reply by GiudGiud** [reply]
same for the auxkernels

**Reply by RECHOA** [reply]
Sorry maybe i have misinformed you. Although the auxkernels are related to the location paramete, but i declared the same auxvariable. 
For example, when calculting the radial displacement, i declared the same "`variable = rad_disp`":
```
  [./raddispaux_73]
    type = RadialDisplacementSphereAux
    variable = rad_disp
    block = '363 '
    origin = '0.000270492 -0.00235077 -0.00270761'
  [../]
  [./raddispaux_74]
    type = RadialDisplacementSphereAux
    variable = rad_disp
    block = '368'
    origin = '-0.00152647 0.000133102 -0.00589633'
  [../]
  ...
```


As for the eigen strian, they are in the form like this:
```
  [./EigenStrain_232]
    type = ComputeMyMaterialEigenstrain
    temperature = temp
    eigenstrain_name = 'pyc_eigen_232'
    block = '1160'
    origin = '-0.000275106 -0.00174505 7.86264e-05'
  [../]
  [./_EigenStrain_233]
    type = ComputeMyMaterialEigenstrain
    temperature = temp
    eigenstrain_name = 'pyc_eigen_233'
    block = '1163 '
    origin = '0.00161647 -0.0019067799999999998 0.00480855'
  [../]
  ...
```

**Reply by GiudGiud** [reply]
Same idea here.
You could use a file for origin rather than create a 1000 objects

**Reply by RECHOA** [reply]
Thanks for your reply. Do you mean creating the Eigenstrain object like following form?
```
  [./EigenStrain_all]
    type = ComputeMyMaterialEigenstrain
    temperature = temp
    eigenstrain_name = 'eigen_all'
    block = all_block
    origin = 'origin_file'  or.  'origin_vector'
  [../]
```
However, different blocks  corresponding to different origins. That is, different blocks  corresponding to different Eigenstrain. If i simply specify `block = all_block`, the seperate block cannot automatically  identify its origin.
So even if using a file or vector for origin, i should write many `"[./EigenStrain_xxx]"` to specify each calculation area (block) and its origin.

**Reply by GiudGiud** [reply]
If you code it you could load an origin vector that is indexed by subdomains. 
since you have 3000 you will want to load it from a file rather than have it in the input file. 


and actually the FileReaderUserObject has indexing by sub domain reading of files already. So most of the code you need is in the PiecewiseConstrantFromCSV function

**Reply by RECHOA** [reply]
Thanks for your relpy. According to my immature expectations, setting materials objects in another way wouldn't affect the mermory becase the mesh number is the same. Acurally, I have tried to run a case with less spheres (the number is about 1/3 of previous case) , it run successfully, and when it went to compute the initial materials, it shows:

>     Setting Up Materials
>       Computing Initial Material Values                                                  [[33m 11.66 s[39m] [[33m 11429 MB[39m]
>     Finished Setting Up Materials                                                        [[33m 11.73 s[39m] [[33m 11429 MB[39m]

which means the memory usage is 11429 MB (11.2GB).  Since the previous case has 3 times of spheres, the memory demand would be about 33.6GB, supposing it increases linearly with the number of mesh. Anyway it wouldn't exceed 2TB theoretically.

**Reply by GiudGiud** [reply]
did you watch the memory during this simulation?
The memory usage reported is a delta between the beginning and end, it would not capture a spike.

**Reply by RECHOA** [reply]
I used the `show_actions = true` and here is the teminal record for the smaller case:
[smallcase_result.txt](https://github.com/idaholab/moose/files/9846972/smallcase_result.txt)

**Reply by GiudGiud** [reply]
It seems like it ran well. Let's look at the sbatch script

**Reply by RECHOA** [reply]
The sbatch script is:
```
#!/bin/bash
#SBATCH -p amd_2T
#SBATCH -N 1
#SBATCH -n 64
source  activate /public1/home/scb3820/anaconda3/envs/moose/
export PATH=/public1/home/scb3820/new-projects/bees_new:$PATH
mpiexec -n 60  ./*opt -i input_master.i
```
Just now, i run the small case again and watched  its memory usage using `watch “free -g”`. However, the result seems to be different from the terminal log from MOOSE:
![1666541555716](https://user-images.githubusercontent.com/33534979/197403397-f4a2ad83-4171-4196-8e74-9b98331b834e.png)
It shows more than 800GB memory  used in the smaller case.

**Reply by GiudGiud** [reply]
I dont think the log catches spikes. It computes deltas between points.

**Reply by RECHOA** [reply]
when using `watch “free -g”`, it checked the momory every 2.0 s,  i noticed that when ruuning the smaller case before computing initial materials, the "used memory" keeps about 100~200GB, but it shows more than 800 GB used when goes to compute initial materials.

**Reply by GiudGiud** [reply]
ok thank you for diagnosing this thoroughly. 
I'm not familiar with this part of the code. 
@lindsayad @friedmud do you know what "computing initial materials" takes a lot of memory with 3000 materials ?
Any easy steps to reduce it?

**Reply by lindsayad** [reply]
I would guess that there are stateful material properties which do indeed use quite a bit of memory because we have to store values at every quadrature point on every element (and side if needed)

**Reply by GiudGiud** [reply]
Are the materials all properly block restricted @RECHOA ?

**Reply by RECHOA** [reply]
@lindsayad Yes, the thermal properties  and mechanical properties  were stored on every element.

@GiudGiud yes, the materials were restricted by block IDs. Acturally, if any block misses material, the code will output error message about that.

**Reply by GiudGiud** [reply]
Are they restricted as much as possible? We want to make sure the materials dont overlap to save storage

**Reply by RECHOA** [reply]
yes. Firstly, properties of different materials were defined in different C++ classes. for example:
```
ThermalPropertiesOfMaterialA.C
MechanicalPropertiesOfMaterialA.C
ThermalPropertiesOfMaterialB.C
PropertiesOfMaterialB.C
...
```
Secondly, when applying the properties to the material, the blocks were restricted  according to their materials.

**Reply by GiudGiud** [reply]
in those different materials class, do you define the minimum number of properties?
when you restricted materials to blocks, was it something like
```
[Materials]
  [thermal_properties]
    type = ...
    block = 23
  []
   ...
[]
```

**Reply by RECHOA** [reply]
Acturally, in `ThermalPropertiesOfMaterialxx.C` I defined thermal conductivity, density and specific heat.
For blocks with the same materials, I restricted them together, and blocks with other materials would be in the other objects.
For example,  'block 23 24 25 26' are material A, 'block 27 28 29 30' are material B, then it will be like
```
[Materials]
  [thermal_properties_A]
    type = ...
    block = '23 24 25 26 '
  []
  [thermal_properties_B]
    type = ...
    block = '27 28 29 30 '
  []
   ...
[]
```

**Reply by GiudGiud** [reply]
ok this seems fine.

@lindsayad when you have stateful materials, does it compute and save ALL materials or just the ones needed for statefulness?

**Reply by lindsayad** [reply]
Only stateful material properties are saved on a per-elem, per-qp basis. There's not really anything like a "stateful" material. A material may provide a diffusivity and then a consumer may request the old value of that diffusivity. This makes the diffusivity stateful irrespective of anything going on in the material declaring the diffusivity

**Reply by RECHOA** [reply]
My original case had consicered the thermal expansion and swelling of materials. I tried to calculate the thermal expansion only, but the memory also exceeded. I find an article simulating the similar object by using code BISON, which is also developed base on MOOSE. 
https://www.sciencedirect.com/science/article/abs/pii/S002954931830325X
and in this paper, it shows :

> The following simulations were performed on the FALCON cluster at INL. The mesh used for the LWR simulations has 2.3 million degrees of freedom. The degrees of freedom (DOF) allocated for the variables and aux-variables was 9.2 and 115.4 million respectively. In these simulations, 256 processors with 1.9 TB of ram were used and the resulting run time was about one day.

It has only  2.3 million degrees of freedom but used a large number of memory (1.9 TB). So maybe it's a general problem in MOOSE?

**Reply by lindsayad** [reply]
@dschwen @bwspenc @recuero could someone comment on whether this type of memory usage is typical for a thermomechanics simulation with stateful properties?

@RECHOA if you could create a MOOSE-only input or share a branch from a public github repository, then we could probably do some memory profiling and look at hot spots

**Reply by RECHOA** [reply]
@lindsayad The mesh is large and complicated, and sorry that I couldn't share a branch due to some commercial reasons of source code. The main structural form of the input file is as follows, it perhaps help：
[inputfile_form.txt](https://github.com/idaholab/moose/files/9858479/inputfile_form.txt)

**Reply by lindsayad** [reply]
Yea due to limited time we cannot debug incomplete inputs

**Reply by RECHOA** [reply]
I created a  MOOSE-only input in which a cylinder was divided into 100 blocks (subdomains). For its eigenstrain, only thermal expansion was considered：
[input_Cylinder_100blocks.txt](https://github.com/idaholab/moose/files/9868842/input_Cylinder_100blocks.txt)

To reflect that different eigenstrains were applied to different blocks in my original input file, `[./thermal_eigenstrain_1] ` to `[./thermal_eigenstrain_100] ` and `[./strain_1]` to `[./strain_100]` were applied in this MOOSE-only input.

The number of all element is about 5 million (close to that in my original input), again I used the command `watch 'free -g'` to see the memory usage and found it increased to more than 1TB when the case run into
>     Setting Up Materials
>       Computing Initial Material Values.....

![memory](https://user-images.githubusercontent.com/33534979/198014306-f71c469b-e451-416d-a46e-cbf4347b2f60.png)

**Reply by hugary1995** [reply]
`ComputeThermalExpansionEigenstrain` doesn't have stateful properties. `ComputeFiniteStrain` does.

**Reply by hugary1995** [reply]
Try to consolidate your `ComputeFiniteStrain` object into one as @GiudGiud previously suggested. Something like this:
```python
    [strain]
      type = ComputeFiniteStrain
      eigenstrain_names = 'thermal_expansion'
    []
```
instead of tons of objects. Give all your block-restricted thermal eigen strains the same property name, instead of appending a different number for each.

**Reply by RECHOA** [reply]
@hugary1995 Do you mean that I should apply `[./thermal_eigenstrain_1] `to `[./thermal_eigenstrain_100] ` the same `eigenstrain_name` like this following form? 
```
    [./thermal_eigenstrain_1]
      type = ComputeThermalExpansionEigenstrain
      eigenstrain_name = 'thermal_expansion'
      block = '1 '
    [../]
    [./thermal_eigenstrain_2]
      type = ComputeThermalExpansionEigenstrain
      eigenstrain_name = 'thermal_expansion'
      block = '2 '
    [../]
    ......
    [./strain]
      type = ComputeFiniteStrain
      eigenstrain_names = 'thermal_expansion'
      block = '1 2 3 4 5 ........ 100 '
    [../]
```

**Reply by hugary1995** [reply]
Here's what I tested:

1. I reduced your problem size by setting `num_sectors = 6`, reducing the number of rings to 10 and number of layers to 10. So that the problem is small enough. Therefore, there are only 10 blocks, 10 thermal eigenstrains, 10 strain calculators in the input file.
2. With 10 strain calculators, the memory consumed by `ReferenceResidualProblem::computingInitialStatefulProps` is ~230 MiB.
3. If I consolidate 10 strain calculators into 1, the memory consumption reduces to ~80 MiB. I expect the reduction to be more pronounced in your case as you have many more strain calculators.

**Reply by hugary1995** [reply]
> Do you mean that I should apply [./thermal_eigenstrain_1] to [./thermal_eigenstrain_100] the same eigenstrain_name like this following form?

Yes. You can also omit the `block = '1 2 3 4 5 ........ 100 '` thing, as the default is all.

**Reply by hugary1995** [reply]
@GiudGiud @lindsayad I'm still interested in understanding why this (the memory consumption grows exponentially with number of objects, although the stateful properties are block-restricted) happens though.

**Reply by RECHOA** [reply]
@hugary1995 Thanks for your test. Also I run the  MOOSE-only input again after reducing strain calculator into 1 as you advised, now the max memory usage is about 500GB, which is about half of previous one. So the memory usage is influenced by the number of strain objects as you and @GiudGiud expected. I am also interested in why this happends. Before, I thought the calculation amount would not changed, since the number of element is the same.

**Reply by lindsayad** [reply]
@hugary1995 can you paste your modified input please?

**Comment by dschwen** [comment]
Can you show your PBS script?

**Reply by RECHOA** [reply]
The script is like this:
```
#!/bin/bash
#SBATCH -p amd_2T
#SBATCH -N 1
#SBATCH -n 64
source  activate /public1/home/scb3820/anaconda3/envs/moose/
export PATH=/public1/home/scb3820/new-projects/bees_new:$PATH
mpiexec -n 60  ./*opt -i input_master.i
```

**Reply by GiudGiud** [reply]
do you need to request memory?
for example
#SBATCH --mem=2G           # total memory per node
